### ISSUE TITLE
Doesn't build on OS X because bison is too old

### ISSUE THREAD

**Issue by msabramo** [mwilliamson/jq.py#4]
```
$ pip install jq
...
    checking bison version... no
    configure: error: You need bison version 2.5 or more recent.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/Users/marca/dev/git-repos/jq.py/setup.py", line 77, in <module>
        'Programming Language :: Python :: 3.4',
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/core.py", line 152, in setup
        dist.run_commands()
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/Users/marca/python/virtualenvs/jq.py/lib/python2.7/site-packages/setuptools/command/develop.py", line 32, in run
        self.install_for_development()
      File "/Users/marca/python/virtualenvs/jq.py/lib/python2.7/site-packages/setuptools/command/develop.py", line 117, in install_for_development
        self.run_command('build_ext')
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/Users/marca/dev/git-repos/jq.py/setup.py", line 41, in run
        command(["./configure", "CFLAGS=-fPIC"])
      File "/Users/marca/dev/git-repos/jq.py/setup.py", line 38, in command
        subprocess.check_call(args, cwd=jq_lib_dir)
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 540, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['./configure', 'CFLAGS=-fPIC']' returned non-zero exit status 1
    Complete output from command /Users/marca/python/virtualenvs/jq.py/bin/python -c "import setuptools, tokenize; __file__='/Users/marca/dev/git-repos/jq.py/setup.py'; exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" develop --no-deps:
```

I'm trying to figure out if I can coax it into building...

**Comment by msabramo** [comment]
I have a newer bison installed with Homebrew but it's not on the PATH.

I think I've gotten it to use the newer bison by doing:

```console
YACC=/usr/local/Cellar/bison/3.0.3/bin/yacc pip install jq
```

but then I get:

```console
...
    ld: targeted OS version does not support use of thread local variables in _jv_nomem_handler for architecture x86_64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    error: command 'gcc' failed with exit status 1
    Complete output from command /Users/marca/python/virtualenvs/jq.py/bin/python -c "import setuptools, tokenize; __file__='/Users/marca/dev/git-repos/jq.py/setup.py'; exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" develop --no-deps:
...
```

**Comment by mwilliamson** [comment]
Do you get that error when jq is being built, or the Python bindings? If it's jq, then it's probably best to try to build jq 1.4 separately to see if you get any errors. If you do, then the jq maintainers themselves can probably help more than I can.

**Comment by msabramo** [comment]
It's when building the Python bindings - let me share more of the output:

```
...
    building 'jq' extension
    gcc -fno-strict-aliasing -fno-common -dynamic -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/marca/dev/git-repos/jq.py/jq-jq-1.4 -I/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c jq.c -o build/temp.macosx-10.4-x86_64-2.7/jq.o
    jq.c:1242:104: warning: passing 'const char *' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
        ((struct __pyx_vtabstruct_2jq__ErrorStore *)__pyx_v_store->__pyx_vtab)->store_error(__pyx_v_store, jv_string_value(__pyx_v_error));
                                                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
    jq.c:4579:42: warning: function '__Pyx_PyInt_AsSignedLongLong' is not needed and will not be emitted [-Wunneeded-internal-declaration]
    static CYTHON_INLINE signed PY_LONG_LONG __Pyx_PyInt_AsSignedLongLong(PyObject* x) {
                                             ^
    22 warnings generated.
    gcc -bundle -undefined dynamic_lookup build/temp.macosx-10.4-x86_64-2.7/jq.o /Users/marca/dev/git-repos/jq.py/jq-jq-1.4/.libs/libjq.a -o /Users/marca/dev/git-repos/jq.py/jq.so
    ld: targeted OS version does not support use of thread local variables in _jv_nomem_handler for architecture x86_64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    error: command 'gcc' failed with exit status 1
    Complete output from command /Users/marca/python/virtualenvs/jq.py/bin/python -c "import setuptools, tokenize; __file__='/Users/marca/dev/git-repos/jq.py/setup.py'; exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" develop --no-deps:
...
```

`jq` itself has already built OK:

```
[marca@marca-mac2 jq.py]$ ll jq-jq-1.4/jq
-rwxr-xr-x+ 1 marca  staff   274K Jan 16 19:09 jq-jq-1.4/jq*
[marca@marca-mac2 jq.py]$ file jq-jq-1.4/jq
jq-jq-1.4/jq: Mach-O 64-bit executable x86_64
```

**Comment by msabramo** [comment]
Note to self: A nicer way to reproduce this that result in less junky output is:

```
$ YACC=/usr/local/Cellar/bison/3.0.3/bin/yacc python setup.py build
...
creating build/lib.macosx-10.4-x86_64-2.7
gcc -bundle -undefined dynamic_lookup build/temp.macosx-10.4-x86_64-2.7/jq.o jq-jq-1.4/.libs/libjq.a -o build/lib.macosx-10.4-x86_64-2.7/jq.so
ld: targeted OS version does not support use of thread local variables in _jv_nomem_handler for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
error: command 'gcc' failed with exit status 1
```

**Comment by msabramo** [comment]
I wondered if it had to do with the fact that it was using `gcc` and recent OS X versions prefer `clang`.

```
$ YACC=/usr/local/Cellar/bison/3.0.3/bin/yacc CC=clang python setup.py build
...
clang -bundle -undefined dynamic_lookup build/temp.macosx-10.4-x86_64-2.7/jq.o jq-jq-1.4/.libs/libjq.a -o build/lib.macosx-10.4-x86_64-2.7/jq.so
ld: targeted OS version does not support use of thread local variables in _jv_nomem_handler for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
error: command 'clang' failed with exit status 1
```

No dice.

**Comment by msabramo** [comment]
Huh? It works when I do the `clang` link command manually from the shell; it just doesn't work within `python setup.py build`?

```
[marca@marca-mac2 jq.py]$ clang -bundle -undefined dynamic_lookup build/temp.macosx-10.4-x86_64-2.7/jq.o jq-jq-1.4/.libs/libjq.a -o build/lib.macosx-10.4-x86_64-2.7/jq.so
[marca@marca-mac2 jq.py]$ echo $?
0
[marca@marca-mac2 jq.py]$ ls -l build/lib.macosx-10.4-x86_64-2.7/jq.so
-rwxr-xr-x+ 1 marca  staff  304424 Jan 17 09:26 build/lib.macosx-10.4-x86_64-2.7/jq.so*
```

**Comment by msabramo** [comment]
I wonder if this has something to do with `MACOSX_DEPLOYMENT_TARGET` ?

**Comment by msabramo** [comment]
Ooh. The following appears to work:

```
MACOSX_DEPLOYMENT_TARGET="10.5" YACC=/usr/local/Cellar/bison/3.0.3/bin/yacc CC=clang python setup.py build
```

Later on, I'll try to formulate this into a PR so that it's automatic. Or failing that I'll send a PR to add a note to the README.

**Comment by mwilliamson** [comment]
Thanks for investigating! I'm not at all familiar with building on a Mac: do you know why the default deployment target doesn't work?

**Comment by msabramo** [comment]
Good question. I'm not too familiar with OS X build stuff easier but I found this in the spidermonkey project:

~~~python
try:
    # For some reason, importing Pyrex appears to change the value of
    # the MACOSX_DEPLOYMENT_TARGET to something that may be
    # inaccurate.  So we'll make sure that it's set to whatever it was
    # before we imported Pyrex.
    oldTarget = os.environ.get("MACOSX_DEPLOYMENT_TARGET", "")

    from Pyrex.Distutils import build_ext

    os.environ["MACOSX_DEPLOYMENT_TARGET"] = oldTarget
~~~

I'll play with this later.

**Comment by msabramo** [comment]
[Docs on MACOSX_DEPLOYMENT_TARGET](https://developer.apple.com/library/mac/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW129)

**Comment by msabramo** [comment]
This likely depends on what Python installation one is using and on what version of OS X (as they come with different Python versions in `/usr/bin/python`.

For example, on my OS X 10.9.5 system:

```
[marca@marca-mac2 ~]$ which python
/Library/Frameworks/Python.framework/Versions/2.7/bin/python
[marca@marca-mac2 ~]$ /Library/Frameworks/Python.framework/Versions/2.7/bin/python -c 'import sysconfig; print(sysconfig.get_config_var("MACOSX_DEPLOYMENT_TARGET"))'
10.4
[marca@marca-mac2 ~]$ /usr/bin/python -c 'import sysconfig; print(sysconfig.get_config_var("MACOSX_DEPLOYMENT_TARGET"))'
10.9
```

**Comment by msabramo** [comment]
I upgraded from Python 2.7.6 to Python 2.7.9 and now I have:

```
[marca@marca-mac2 ~]$ /Library/Frameworks/Python.framework/Versions/2.7/bin/python -c 'import sysconfig; print(sysconfig.get_config_var("MACOSX_DEPLOYMENT_TARGET"))'
10.6
```

**Comment by mwilliamson** [comment]
Thanks for the pull requests! I've merged them in and released version 0.1.2 to PyPI, which should hopefully fix the issues.

**Comment by msabramo** [comment]
Looks great! Thank you!

```
❯ workon jq.py
...
❯ pip install jq
Collecting jq
  Downloading jq-0.1.2.tar.gz
Installing collected packages: jq
...
    22 warnings generated.
    /usr/bin/clang -bundle -undefined dynamic_lookup -arch i386 -arch x86_64 -g build/temp.macosx-10.6-intel-2.7/jq.o /var/folders/gw/w0clrs515zx9x_55zgtpv4mm0000gp/T/pip-build-RTXOyh/jq/jq-jq-1.4/.libs/libjq.a -o build/lib.macosx-10.6-intel-2.7/jq.so
    ld: warning: ignoring file /var/folders/gw/w0clrs515zx9x_55zgtpv4mm0000gp/T/pip-build-RTXOyh/jq/jq-jq-1.4/.libs/libjq.a, file was built for archive which is not the architecture being linked (i386): /var/folders/gw/w0clrs515zx9x_55zgtpv4mm0000gp/T/pip-build-RTXOyh/jq/jq-jq-1.4/.libs/libjq.a
Successfully installed jq-0.1.2

❯ ipython --no-banner

In [1]: from jq import jq

In [2]: jq(".[]+1").transform([1, 2, 3], multiple_output=True)
Out[2]: [2, 3, 4]
```

Thanks!
